include/lbann/trainers/trainer.hpp

```diff
@@ -77,6 +77,18 @@ class trainer {
   /** Human-readable description. */
   description get_description() const;
 
+  /** @brief Get the list of callbacks for the trainer. */
+  std::vector<std::shared_ptr<callback_base>>& get_callbacks() {
+    return m_callbacks;
+  }
```

Comment:
I both agree and disagree. I don't like this because it exposes the ownership structure, but I disagree that it "gives ownership" to the caller -- ownership is tied to the reference count, which will not go up here. The tradeoff seems expensive to me for little benefit. In some sense, though, this will hopefully be moot after refactoring the callbacks.

Reply:
Good point that it returns a reference, not a copy. In any case, I don't think we'd lose much performance by copying the shared_ptrs into raw pointers. This isn't happening in an inner loop, so the gains in clarity and safety are worth the cost of allocating some memory on the heap.